Detect HTLC-resolving on-chain actions and pass them to ChannelManager

TheBlueMatt opened the pull request:
This is #198 rebased and with a number of substantial and many non-substantial changes. The full diff against a simple rebase on master is available at the 0bin link, but some changes include:
 * failing backwards HTLCs after a remote revoked commitment tx broadcast based on last two remote commitment tx HTLC contents, instead of last two local commitment tx HTLC contents (with a new test showing why this is required),
 * move updating of HTLCs in ChannelManager into the event getter instead of block_connected, reducing a bunch of duplicates, although this is also a shitty situation, I think its a bit better and we can resolve it with #80.

https://0bin.net/paste/icbkyeuL3WD7nhVA#tLgGthEsvsid0wHky0SKVW3-0+TpTmru409h/X6A/sr

TheBlueMatt commented on src/ln/channelmanager.rs:
```diff
@@ -2622,6 +2622,20 @@ impl ChannelManager {
+				if let Some(preimage) = htlc_update.payment_preimage {
+					self.claim_funds_internal(self.channel_state.lock().unwrap(), htlc_update.source, preimage);
+				} else {
+					self.fail_htlc_backwards_internal(self.channel_state.lock().unwrap(), htlc_update.source, &htlc_update.payment_hash, HTLCFailReason::Reason { failure_code: 0x4000 | 10, data: Vec::new() });
```
Ah, hmm, can we fix this in a separate PR cause it needs to be fixed in a few other places too, I think.

ariard replied:
Ok, will open a separate PR so

TheBlueMatt commented on src/ln/channelmanager.rs:
```diff
@@ -6519,6 +6519,224 @@ mod tests {
+
+				if !deliver_bs_raa {
+					// If we delievered B's RAA we got an unknown preimage error, not something
+					// that we should update our rouring table for.
```
heh, fixed.

TheBlueMatt commented on src/ln/channelmanager.rs:
```diff
@@ -6519,6 +6519,224 @@ mod tests {
+		}
+		if deliver_bs_raa {
+			match events[0] {
+				MessageSendEvent::UpdateHTLCs { ref node_id, updates: _ } => {
```
checked.

TheBlueMatt commented on src/ln/channelmanager.rs:
```diff
@@ -2622,6 +2622,20 @@ impl ChannelManager {
 
 impl events::MessageSendEventsProvider for ChannelManager {
 	fn get_and_clear_pending_msg_events(&self) -> Vec<events::MessageSendEvent> {
+		// TODO: Event release to users and serialization is currently race-y: its very easy for a
```
Well any of the actual results are going to need to wait on an event clear anyway. eg if it results in sending an update_fulfill_htlc we still have to wait for them to call get_and_clear_pending_msg_events. It may even make sense to move the transaction broadcasting stuff into events as the whole broadcaster API is pretty awkward.

As for the race condition here - either we can serialize the pending events or we can just fail serialization if there are pending events and let the user go in a loop until there are none left, which I think is probably ok.

ariard replied:
Ah yes, so I think it's fine, acknowledging it's a best effort for now, need to spend a little bit more time with rust-lightning-bitcoinrpc to have more clues on how the broadcast/events APIs could be used

TheBlueMatt commented on src/ln/channelmonitor.rs:
```diff
@@ -122,27 +134,63 @@ pub struct SimpleManyChannelMonitor<Key> {
+						let mut existing_claim = false;
+						e.get_mut().retain(|htlc_data| {
+							if htlc.0 == htlc_data.0 {
+								if htlc_data.1.is_some() {
```
Well we definitely cant want for 100 blocks cause we have to fail backwards before then. For now I think 6 is probably fine. I added the constant for it in a new commit with appropriate checking.

ariard replied:
Good, do you push the commit on your local branch, head commit is fb29935 currently ?

TheBlueMatt commented:
Pushed a big chunk rewrite as we need to consider not-included-in-commitment-transaction-due-to-dust-limit HTLCs as well. This has significant implications for the design layout here, so I went ahead and fixed those, though I did not actually go implement all the bits that will be needed for correctness. I added TODOs in a few places noting this, however.

TheBlueMatt commented:
Also further loosened the debug_assert!()s in Channel get_update_*_htlc functions, since they were still too strict and fuzzer was finding ways to hit them.

TheBlueMatt commented on src/ln/channelmonitor.rs:
```diff
@@ -1617,6 +1692,64 @@ impl ChannelMonitor {
+					if payment_data.is_none() {
+						for htlc_output in $htlc_outputs {
+							if input.previous_output.vout == htlc_output.transaction_output_index {
+								log_info!(self, "Inbound HTLC timeout at {} from {} resolved by {}", input.previous_output.vout, input.previous_output.txid, tx.txid());
```
Rewrote it to be closer to the above one.

TheBlueMatt commented on src/ln/channelmonitor.rs:
```diff
@@ -1407,6 +1407,10 @@ impl ChannelMonitor {
 					});
 					txn_to_broadcast.push(spend_tx);
+
+					// TODO: We need to fail back HTLCs that were't included in the broadcast
```
If we create a transaction for the remote side, we *must* provide_latest_remote_tx (and update the monitor, and write it all to disk) before we send the signatures required for the remote to broadcast it. For local, we either have to succeed at committing new monitors or we have to fail the channel immediately using the current in-memory monitor (though I think we need to move to failing HTLCs based only on remotes, so this wont matter anyway).

ariard replied:
To clarify my question, what if remote peer broadcast its commitment tx while we're done with inbounds htlcs (so we have them in HTLCForwardInfo after Channel::forward_htlcs) but before we call process_pending_htlc_forwards ? I didn't test it, but seems we're unable to fail backward inbound HTLCs...

TheBlueMatt replied:
Then the counterparty would be failing a transaction with inbound HTLCs, not outbound ones, which we aren't really handling in this patch. If they dutifully committed to an HTLC (implying the corresponding ChannelMonitor was updated) and it ended up in our pending HTLC forwards there is nothing wrong with forwarding it, it just means when the upstream node either fails or forwards we need to get that information back to the original ChannelMonitor so it can do the same on chain. Not forwarding it may be nicer since we could avoid the on-chain fee and just let them claim the timeout, but that is a relatively small edge case and adding code to handle it probably isnt worth it.

ariard replied:
Okay good, well letting other pay the fee would be a nice extension of what a greedy algorithm means.

TheBlueMatt commented on src/ln/channelmonitor.rs:
```diff
@@ -1746,6 +1753,16 @@ impl ChannelMonitor {
+		// transaction (or in any of the latest two local commitment transactions). This probably
+		// needs to use the same logic as the revoked-tx-announe logic - checking the last two
+		// remote commitment transactions. This probably has implications for what data we need to
+		// store in local commitment transactions.
```
I think we may be able to store *less* data.

ariard replied:
So the idea is to fail backward htlc at would_broadcast_height so we don't have to extract timeout updates from txs spending our own commitment tx  ? Would be smarter effectively, seems we need to put stuff too in ChannelManager::force_close_channel

TheBlueMatt replied:
Nono, I think we need to switch from examining the latest local commitment tx here to examining the latest two remote commitment transactions like we do in check_spend_remote_transaction in case of revoked tx broadcast. This implies we may be able to remove some of the HTLC info from local txn since we won't be using that here.

ariard replied:
Hmm let accept we do that, so we would generate htlc_updated in would_broadcast_height based on what is inside latest two remote commitment tx ? If so we would pass backward timeout htlc on a not-yet-broadcast local commitment tx does it raise any problem ? Doing this would effectively mean less stuff in latest local commitment tx and simplify is_resolving_htlc_output

TheBlueMatt replied:
Hmm, to be honest I haven't fully thought through this one, hence why I just slapped on the TODO and moved forward (this thing really needs to get merged...). It may turn out we can't do something so simple, but intuitively it seems like we may be ignoring some HTLCs here where we shouldn't be.

TheBlueMatt commented on src/ln/channelmonitor.rs:
```diff
@@ -1746,6 +1753,16 @@ impl ChannelMonitor {
+		// TODO: We need to consider HTLCs which were below dust threshold here - while they don't
+		// strictly imply that we need to fail the channel, we need to go ahead and fail them back
+		// to the source, and if we don't fail the channel we will have to ensure that the next
+		// updates that peer sends us are update_fails, failing the channel if not. It's probably
```
Indeed, there would have to be a bit of plumbing of stuff from monitors to Channels themselves, so I agree, its probably easier to not worry about this edge case and just always fail to chain.

ariard replied:
Agree we don't have to broadcast commitment tx because we can't timeout onchain these trimmed HTLCs, there are not presents in the commitment tx but protocol logic ask us to fail backward and ideally we should wait for remote update_fail but we go ahead to avoid layers violation, Maybe remove the TODO and just let a comment?

TheBlueMatt replied:
The TODO is that we need to do *something* - I think the right something is to just close the channel wholesale if a below-dust HTLC times out, but we need to switch to doing that from the current code which still only looks at the HTLCs present in the commitment tx.

ariard replied:
Spoke too fast, yes

ariard commented on src/ln/channel.rs:
```diff
@@ -1179,15 +1191,14 @@ impl Channel {
-							// Return the new channel monitor in a last-ditch effort to hit the
-							// chain and claim the funds
+							log_warn!(self, "Have preimage and want to fulfill HTLC with pending failure against channel {}", log_bytes!(self.channel_id()));
+							// TODO: We may actually be able to switch to a fulfill here, though its
```
Hmm not sure that the remote sending us or not its RAA changes anything, just him forwarding us update_fulfill_htlc later than the safe delay-fail-backwards mean we lost opportunity to claim HTLC backward. Agree it should only happen with reorg > delay-fail-backward, so rare enough but maybe fun to write a test on it, will wait to write delay-fail-backward stuff first :)

TheBlueMatt replied:
This case is only the holding cell, so we can only get here if we went to fail backwards initially and couldn't as we were waiting on an RAA. If we weren't waiting on an RAA we would have failed backwards immediately and have no recourse if a reorg happens.

ariard replied:
Ah yes ! Was thinking quickly we would only fail backward after reorg delay exhaustion, but that would slow down the network so much so really dumb, doesn't make sense

TheBlueMatt replied:
Oops, I think we're talking past each other a bit, too many different channels and things involved...no, I think we should delay failing backwards until reorg delay exhaustion...once things have hit the chain we're gonna be mega slow compared to regular operation anyway, being slower still isn't much of a hit.

The case here is where an upstream channel hit the chain and, after it claimed an HTLC with a timeout, we waited a few blocks to make sure there wasn't gonna be a reorg, then failed it backwards. When we went to fail it backwards, we found that the downstream channel was in AwaitingRAA (ie it hasn't send us an RAA after we sent it something), so we added the failure to the holding cell. Then, we actually got a really surprising reorg and suddenly had a claim on the original, upstream, on-chain output, resulting in a new fulfill event! When that event came through we found that the downstream channel *still* hadn't provided the needed RAA so we got here.

Phew...that was a lot of rare cases, so probably not worth special (at least untested) code to handle :P.
